tests/lms/models/group_info_test.py

```diff
@@ -0,0 +1,82 @@
+import pytest
+from sqlalchemy.exc import IntegrityError
+
+from lms.models import ApplicationInstance, GroupInfo
+
+
+class TestGroupInfo:
+    def test_persist_and_retrieve(self, application_instance, db_session, group_info):
+        db_session.add(group_info)
+
+        retrieved_group_info = db_session.query(GroupInfo).one()
+
+        assert retrieved_group_info.id
+        assert (
+            retrieved_group_info.authority_provided_id == "test_authority_provided_id"
+        )
+        assert retrieved_group_info.consumer_key == "test_consumer_key"
+
+    def test_application_instance_relation(
+        self, application_instance, db_session, group_info
+    ):
+        db_session.add(group_info)
+
+        retrieved_group_info = db_session.query(GroupInfo).one()
```

Comment:
I don't know enough about SQLAlchemy to know if this is making a full round trip to the DB or just in memory. If it's not I think we'd definitely want to make sure it was. Does that mean a flush before the query or do I just not get this?

Reply:
The query will cause the SQL to be autoflushed to Postgres. One way to test this is to turn on debug logging of the SQL statements that get sent:

```diff
diff --git a/tests/lms/conftest.py b/tests/lms/conftest.py
index 943baa34..6cc66fb3 100644
--- a/tests/lms/conftest.py
+++ b/tests/lms/conftest.py
@@ -28,7 +28,7 @@ SESSION = sessionmaker()
 
 @pytest.fixture(scope="session")
 def db_engine():
-    engine = sqlalchemy.create_engine(TEST_DATABASE_URL)
+    engine = sqlalchemy.create_engine(TEST_DATABASE_URL, echo=True)
     db.init(engine)
     return engine
 ```

Then tell pytest not to swallow the stdout/logging from tests that pass:

```terminal
$ export PYTEST_ADDOPTS="-s"
```

Then run just the one test:

```terminal
$ tox tests/lms/services/group_info_upsert_test.py::TestGroupInfoUpsert::test_it_adds_a_new_GroupInfo_if_none_exists
```

You should see a lot of SQL statements printed, including one `INSERT into application_instances ...` (the `application_instance` fixture) and four `INSERT into group_info`'s (the three from the `group_infos` noise fixture and the fourth from the call to the upsert service in the test body itself).

If you set `engine="debug"` instead of `True` it'll also print out the results of SQL queries and you can see the `'TEST_AUTHORITY_PROVIDED_ID'` etc coming back out of the DB.